src/frequenz/sdk/timeseries/_quantities.py

```diff
@@ -0,0 +1,686 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""Types for holding quantities with units."""
+
+from __future__ import annotations
+
+import math
+from datetime import timedelta
+from typing import Self, TypeVar, overload
+
+QuantityT = TypeVar("QuantityT", "Quantity", "Power", "Current", "Voltage", "Energy")
+
+
+class Quantity:
+    """A quantity with a unit."""
+
+    _base_value: float
+    """The value of this quantity in the base unit."""
+
+    _exponent_unit_map: dict[int, str] | None = None
+    """A mapping from the exponent of the base unit to the unit symbol.
+
+    If None, this quantity has no unit.  None is possible only when using the base
+    class.  Sub-classes must define this.
+    """
+
+    def __init__(self, value: float, exponent: int = 0) -> None:
+        """Initialize a new quantity.
+
+        Args:
+            value: The value of this quantity in a given exponent of the base unit.
+            exponent: The exponent of the base unit the given value is in.
+        """
+        self._base_value = value * 10**exponent
+
+    def __init_subclass__(cls, exponent_unit_map: dict[int, str]) -> None:
+        """Initialize a new subclass of Quantity.
+
+        Args:
+            exponent_unit_map: A mapping from the exponent of the base unit to the unit
+                symbol.
+
+        Raises:
+            TypeError: If the given exponent_unit_map is not a dict.
+            ValueError: If the given exponent_unit_map does not contain a base unit
+                (exponent 0).
+        """
+        if not 0 in exponent_unit_map:
+            raise ValueError("Expected a base unit for the type (for exponent 0)")
+        cls._exponent_unit_map = exponent_unit_map
+        super().__init_subclass__()
+
+    @property
+    def base_value(self) -> float:
+        """Return the value of this quantity in the base unit.
+
+        Returns:
+            The value of this quantity in the base unit.
+        """
+        return self._base_value
```

Comment:
> II think people will feel much more discouraged to use `as_unitless` than base_value in concrete types with units.

how about `as_base_unit` instead?  Because `unitless` is not clear that there's no transformations going on.

> Or maybe we can just override the property base_value with something that raises for types with a unit?

this is not possible, because then we won't be able to have all quantities decompose into the base values.  Then formula engines won't be able to get the base value from Power values, for example.  Or in the feature extractor.

> Or maybe still have a ACB BaseQuantity and have a UnitlessQuantity type 

No, we want any kind of objects to be de-composable to the unitless quantity.  If Power and UnitlessQuantity are in the same level in the hierarchy, then I can't have a type that accepts UnitlessQuantity and accept Power values.  More specific case here:  https://github.com/frequenz-floss/frequenz-sdk-python/pull/422#discussion_r1232042848